core/src/token_info/cached.rs

```diff
@@ -141,6 +154,24 @@ impl TokenInfoFetching for TokenInfoCache {
     fn all_ids<'a>(&'a self) -> BoxFuture<'a, Result<Vec<TokenId>>> {
         self.inner.all_ids()
     }
+
+    fn find_token_by_symbol<'a>(
+        &'a self,
+        symbol: &'a str,
+    ) -> BoxFuture<'a, Result<Option<(TokenId, TokenBaseInfo)>>> {
+        async move {
+            // NOTE: First check the cache directly before refetching. This
+            // allows us to exit early without checking to see if there are
+            // missing token infos that we need to cache.
+            if let Some(result) = self.find_cached_token_by_symbol(symbol).await {
```

Comment:
Are we guaranteed that the cache will be updated in order or at least atomically? Otherwise we might get different results at different times of the request (admittedly a very rare race condition, but also one that seems impossible to debug).

Reply:
> I'd be fine with killing the old paginated orderbook reading

I think we do a `cache_all` on startup, but we can guarantee it.

Reply:
Hmm, yes this is a tough one...

Reply:
I modified the implementation to ensure that all tokens up to the found token are cached. This makes it so the results are reproducible even as new tokens get added to the exchange while still having a "fast-path" so that we don't query the EVM once the cache is sufficiently warmed up.